hist

```diff
@@ -38,26 +47,86 @@ preflight_check() {
     fi
 }
 
+export_historianrc_variables() {
+    # Get custom-set metadata variables with some variable
+    # enumeration hackery from .historianrc (interpreted as a bash script).
+    #
+    # Execute in subshell to avoid interference with the main script.
+    (
+        # This is kind of like a portable env -0 with diff as env -0 is GNU-only.
+        compgen -v > "$tmp"/vars1
+        source "$HISTORIAN_RC"
+        compgen -v > "$tmp"/vars2
+
+        while IFS= read -r var; do
+            echo -ne "${var}=${!var}"'\0'
+        done < <(comm -13 "$tmp/vars1" "$tmp/vars2")
+    )
+    return $?
+}
+
+insert_import_metadata() {
+    # NB. Metadata set here are passed to the DB as-is.
+    # Be careful with quoting.
+    #
+    # In contrast, variables set in historianrc are
+    # passed in fully quoted.
+    #
+    metadata=(
+        imported_at current_timestamp
+    )
+
+    if [[ -f $HISTORIAN_RC ]]; then
+        local -r vars_path=$tmp/historianrc.vars
+
+        if ! export_historianrc_variables > "$vars_path"; then
+            error "Error evaluating historianrc at ${HISTORIAN_RC}: Exiting."
+            exit 10
+        fi
+
+        while IFS='=' read -r -d '' key val; do
+            metadata+=("$key" "'$(sql_quote_term "$val")'")
+        done < "$vars_path"
+    fi
+
+    echo "INSERT into metadata(metadata_id, key, value)"
+    let end=${#metadata[@]}-1
+    for key_idx in $(seq 0 2 $end); do
+        metadata_key=$(sql_quote_term "${metadata[$key_idx]}")
+        # NB. assumed to be appropriately quoted for SQL.
+        metadata_value=${metadata[$key_idx+1]}
+
+        cat <<-EOF
+            $([[ $key_idx == 0 ]] || echo union all)
+            select
+                value as metadata_id,
+                '${metadata_key}' as key,
+                ${metadata_value} as value
+            from variables where key='metadata_id'
+EOF
+    done
+    echo ';'
+}
+
 ensure_db_exists() {
-    ( cat <<SQL
+    historian_sqlite <<SQL
         CREATE TABLE IF NOT EXISTS history (
-            id INTEGER PRIMARY KEY ASC,
-            command TEXT NOT NULL,
-            timestamp INTEGER
+            id INTEGER PRIMARY KEY ASC
+            ,metadata_id INTEGER NOT NULL
+            ,command TEXT NOT NULL
+            ,command_timestamp INTEGER NOT NULL
```

Comment:
Changing the schema here will break existing databases, which complicates things quite a bit, unless there's some sort of schema versioning or migration support.